File: bifrost-rs/crates/bifrost-ejson/src/converters.rs
//! Built-in EJSON converters — port of
//! `bifrost-ts/src/ejson/built-in-converters.ts` and
//! `bifrost-py/src/bifrost/ejson/converters.py`.

use crate::base64::{decode as decode_base64, encode as encode_base64};
use crate::value::{EjsonValue, InfNaNSign};
use chrono::{TimeZone, Utc};
use serde_json::{Map, Number, Value};

/// Wire-faithful regex container (source + flag string).
///
/// JS `RegExp.flags` maps poorly to Rust regex flags, so we preserve
/// the original strings and let callers rebuild a compiled pattern
/// when they want one (same design as the Python port).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Regex {
    pub source: String,
    pub flags: String,
}

const VALID_FLAGS: &[char] = &['g', 'i', 'm', 'u', 'y'];

fn sanitize_flags(raw: &str) -> String {
    let capped = raw.chars().take(50);
    let mut seen = [false; 128];
    let mut out = String::new();
    for c in capped {
        if !VALID_FLAGS.contains(&c) {
            continue;
        }
        let idx = c as usize;
        if idx < seen.len() && !seen[idx] {
            seen[idx] = true;
            out.push(c);
        }
    }
    out
}

/// Convert a Rust `EjsonValue` into its JSON-compatible tag form.
pub fn to_json_value(value: &EjsonValue) -> Value {
    match value {
        EjsonValue::Null => Value::Null,
        EjsonValue::Bool(b) => Value::Bool(*b),
        EjsonValue::Int(i) => Value::Number(Number::from(*i)),
        EjsonValue::Float(f) => Number::from_f64(*f)
            .map(Value::Number)
            .unwrap_or(Value::Null),
        EjsonValue::String(s) => Value::String(s.clone()),

        EjsonValue::Array(items) => {
            Value::Array(items.iter().map(to_json_value).collect())
        }

        EjsonValue::Object(map) => {
            if looks_like_tag_form(map) {
                let mut escape = Map::new();
                for (k, v) in map {
                    escape.insert(k.clone(), to_json_value(v));
                }
                let mut outer = Map::new();
                outer.insert("$escape".to_string(), Value::Object(escape));
                Value::Object(outer)
            } else {
                let mut encoded = Map::new();
                for (k, v) in map {
                    encoded.insert(k.clone(), to_json_value(v));
                }
                Value::Object(encoded)
            }
        }

        EjsonValue::Date(dt) => {
            let millis = dt.timestamp_millis();
            let mut obj = Map::new();
            obj.insert("$date".to_string(), Value::Number(Number::from(millis)));
            Value::Object(obj)
        }

        EjsonValue::Binary(bytes) => {
            let mut obj = Map::new();
            obj.insert(
                "$binary".to_string(),
                Value::String(encode_base64(bytes)),
            );
            Value::Object(obj)
        }

        EjsonValue::Regex(rx) => {
            let mut obj = Map::new();
            obj.insert("$regexp".to_string(), Value::String(rx.source.clone()));
            obj.insert("$flags".to_string(), Value::String(rx.flags.clone()));
            Value::Object(obj)
        }

        EjsonValue::InfNaN(sign) => {
            let mut obj = Map::new();
            obj.insert(
                "$InfNaN".to_string(),
                Value::Number(Number::from(*sign as i64)),
            );
            Value::Object(obj)
        }

        EjsonValue::Custom { name, value } => {
            let mut obj = Map::new();
            obj.insert("$type".to_string(), Value::String(name.clone()));
            obj.insert("$value".to_string(), to_json_value(value));
            Value::Object(obj)
        }
    }
}

/// Inverse of [`to_json_value`].
pub fn from_json_value(value: &Value) -> EjsonValue {
    match value {
        Value::Null => EjsonValue::Null,
        Value::Bool(b) => EjsonValue::Bool(*b),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                EjsonValue::Int(i)
            } else if let Some(f) = n.as_f64() {
                EjsonValue::from_f64(f)
            } else {
                EjsonValue::Null
            }
        }
        Value::String(s) => EjsonValue::String(s.clone()),
        Value::Array(items) => {
            EjsonValue::Array(items.iter().map(from_json_value).collect())
        }
        Value::Object(map) => decode_object(map),
    }
}

fn decode_object(map: &Map<String, Value>) -> EjsonValue {
    // Match the TS gate: tag forms have ≤2 keys and every key starts with '$'.
    if map.len() <= 2 && map.keys().all(|k| k.starts_with('$')) {
        if let Some(decoded) = decode_tag(map) {
            return decoded;
        }
    }
    let mut out = crate::value::EjsonMap::new();
    for (k, v) in map {
        out.insert(k.clone(), from_json_value(v));
    }
    EjsonValue::Object(out)
}

fn decode_tag(map: &Map<String, Value>) -> Option<EjsonValue> {
    if let Some(ms) = map.get("$date").and_then(|v| v.as_i64())
        && map.len() == 1
    {
        return Some(EjsonValue::Date(Utc.timestamp_millis_opt(ms).unwrap()));
    }
    if map.len() == 2
        && let (Some(source), Some(flags)) = (map.get("$regexp"), map.get("$flags"))
        && let (Some(s), Some(f)) = (source.as_str(), flags.as_str())
    {
        return Some(EjsonValue::Regex(Regex {
            source: s.to_string(),
            flags: sanitize_flags(f),
        }));
    }
    if let Some(sign) = map.get("$InfNaN").and_then(|v| v.as_i64())
        && map.len() == 1
    {
        return Some(EjsonValue::InfNaN(match sign {
            0 => InfNaNSign::NaN,
            1 => InfNaNSign::PositiveInfinity,
            _ => InfNaNSign::NegativeInfinity,
        }));
    }
    if let Some(b64) = map.get("$binary").and_then(|v| v.as_str())
        && map.len() == 1
    {
        let bytes = decode_base64(b64).ok()?;
        return Some(EjsonValue::Binary(bytes));
    }
    if let Some(inner) = map.get("$escape").and_then(|v| v.as_object())
        && map.len() == 1
    {
        let mut out = crate::value::EjsonMap::new();
        for (k, v) in inner {
            out.insert(k.clone(), from_json_value(v));
        }
        return Some(EjsonValue::Object(out));
    }
    if map.len() == 2
        && let (Some(name), Some(value)) = (map.get("$type"), map.get("$value"))
        && let Some(name_str) = name.as_str()
    {
        return Some(EjsonValue::Custom {
            name: name_str.to_string(),
            value: Box::new(from_json_value(value)),
        });
    }
    None
}

fn looks_like_tag_form(map: &crate::value::EjsonMap) -> bool {
    if map.len() > 2 || map.is_empty() {
        return false;
    }
    let keys: Vec<&String> = map.keys().collect();
    match (map.len(), keys.first().map(|s| s.as_str())) {
        (1, Some("$date")) => true,
        (1, Some("$InfNaN")) => true,
        (1, Some("$binary")) => true,
        (1, Some("$escape")) => true,
        (2, _) => {
            let ks: Vec<&str> = keys.iter().map(|s| s.as_str()).collect();
            (ks.contains(&"$regexp") && ks.contains(&"$flags"))
                || (ks.contains(&"$type") && ks.contains(&"$value"))
        }
        _ => false,
    }
}

File: typeferry-py/src/typeferry/ejson/converters.py
"""Built-in EJSON converters — port of ``typeferry-ts/src/ejson/built-in-converters.ts``.

The tag forms and order are normative (PROTOCOL.md §3.1):
Date, RegExp, Inf/NaN, Binary, Escape, Custom.

Python type mapping:

* JS ``Date`` ↔ :class:`datetime.datetime` (timezone-aware UTC)
* JS ``RegExp`` ↔ :class:`EJSONRegExp` (source + flags as strings; avoids
  lossy conversion to :class:`re.Pattern`)
* JS ``NaN``/``Infinity``/``-Infinity`` ↔ Python ``float('nan')``,
  ``float('inf')``, ``float('-inf')``
* JS ``Uint8Array`` ↔ Python ``bytes`` / ``bytearray``
* JS custom types ↔ instances implementing :class:`typeferry.ejson.custom_types.CustomType`
"""

from __future__ import annotations

import math
from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

from typeferry.ejson.base64 import decode_base64, encode_base64
from typeferry.ejson.custom_types import (
    get_factory,
    has_type,
    is_custom_type,
)


@dataclass(frozen=True, slots=True)
class EJSONRegExp:
    """Wire-faithful regex container — holds the JS source and flag strings.

    The TS implementation stores ``regexp.source`` (string) and
    ``regexp.flags`` (string like ``"gim"``). Python's :class:`re.Pattern`
    doesn't round-trip these cleanly (JS ``g`` / ``y`` have no Python
    equivalent). Callers who want a compiled pattern should build one
    from :attr:`source` and :attr:`flags` themselves.
    """

    source: str
    flags: str


def _is_inf_or_nan(obj: Any) -> bool:
    if not isinstance(obj, float):
        return False
    return math.isnan(obj) or math.isinf(obj)


def _is_binary(obj: Any) -> bool:
    return isinstance(obj, (bytes, bytearray))


def _has_keys(obj: Any, *keys: str) -> bool:
    if not isinstance(obj, dict):
        return False
    return all(k in obj for k in keys) and len(obj) == len(keys)


# ---------------------------------------------------------------------------
# Converter record
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Converter:
    """Single EJSON tag converter.

    * ``match_json_value(obj)`` — True iff ``obj`` is the JSON-form of this tag
    * ``match_object(obj)`` — True iff ``obj`` is a Python object this converter owns
    * ``to_json_value(obj)`` — serialize the Python object to its tag form
    * ``from_json_value(obj)`` — deserialize the tag form back to a Python object
    """

    match_json_value: Callable[[Any], bool]
    match_object: Callable[[Any], bool]
    to_json_value: Callable[[Any], Any]
    from_json_value: Callable[[Any], Any]


# ---------------------------------------------------------------------------
# Individual converters
# ---------------------------------------------------------------------------


def _date_to_json(obj: datetime) -> dict[str, int]:
    """Encode a datetime as ``{"$date": <ms since epoch>}``.

    Naive datetimes are treated as UTC (matching JS ``Date`` semantics,
    which has no timezone concept beyond the underlying epoch ms).
    """

    epoch = obj.replace(tzinfo=UTC).timestamp() if obj.tzinfo is None else obj.timestamp()
    return {"$date": round(epoch * 1000)}


def _date_from_json(obj: dict[str, Any]) -> datetime:
    ms = obj["$date"]
    return datetime.fromtimestamp(ms / 1000, tz=UTC)


_DATE_CONVERTER = Converter(
    match_json_value=lambda o: _has_keys(o, "$date"),
    match_object=lambda o: isinstance(o, datetime),
    to_json_value=_date_to_json,
    from_json_value=_date_from_json,
)


_VALID_FLAG_CHARS = set("gimuy")


def _sanitize_flags(raw: str) -> str:
    """Match the TS ``fromJSONValue`` flag-sanitization: cap 50 chars,
    filter to ``[gimuy]``, deduplicate."""

    if not isinstance(raw, str):
        raw = ""
    capped = raw[:50]
    filtered = [c for c in capped if c in _VALID_FLAG_CHARS]
    seen: set[str] = set()
    deduped = []
    for c in filtered:
        if c not in seen:
            seen.add(c)
            deduped.append(c)
    return "".join(deduped)


_REGEXP_CONVERTER = Converter(
    match_json_value=lambda o: _has_keys(o, "$regexp", "$flags"),
    match_object=lambda o: isinstance(o, EJSONRegExp),
    to_json_value=lambda r: {"$regexp": r.source, "$flags": r.flags},
    from_json_value=lambda o: EJSONRegExp(
        source=o["$regexp"], flags=_sanitize_flags(o["$flags"])
    ),
)


def _infnan_to_json(obj: float) -> dict[str, int]:
    if math.isnan(obj):
        sign = 0
    elif obj == math.inf:
        sign = 1
    else:
        sign = -1
    return {"$InfNaN": sign}


def _infnan_from_json(obj: dict[str, int]) -> float:
    sign = obj["$InfNaN"]
    if sign == 0:
        return float("nan")
    if sign == 1:
        return float("inf")
    return float("-inf")


_INFNAN_CONVERTER = Converter(
    match_json_value=lambda o: _has_keys(o, "$InfNaN"),
    match_object=_is_inf_or_nan,
    to_json_value=_infnan_to_json,
    from_json_value=_infnan_from_json,
)


_BINARY_CONVERTER = Converter(
    match_json_value=lambda o: _has_keys(o, "$binary"),
    match_object=_is_binary,
    to_json_value=lambda b: {"$binary": encode_base64(bytes(b))},
    from_json_value=lambda o: decode_base64(o["$binary"]),
)


# Forward-declared so ``_ESCAPE_CONVERTER.match_object`` can reference the
# list during recursion.
builtin_converters: list[Converter] = []


def _escape_match_object(obj: Any) -> bool:
    """True iff ``obj`` is a plain dict that looks like a tag form.

    Matches the TS ``matchObject`` for ``$escape``: 1-2 keys AND some
    converter's ``match_json_value`` returns True for it.
    """

    if not isinstance(obj, dict):
        return False
    key_count = len(obj)
    if key_count not in (1, 2):
        return False
    return any(c.match_json_value(obj) for c in builtin_converters)


def _escape_to_json(obj: dict[str, Any]) -> dict[str, Any]:
    # Import lazily — ``to_json_value`` lives in a sibling module that
    # depends on this one.
    from typeferry.ejson.to_from_json_value import to_json_value

    return {"$escape": {k: to_json_value(v) for k, v in obj.items()}}


def _escape_from_json(obj: dict[str, Any]) -> dict[str, Any]:
    from typeferry.ejson.to_from_json_value import from_json_value

    inner = obj["$escape"]
    return {k: from_json_value(v) for k, v in inner.items()}


_ESCAPE_CONVERTER = Converter(
    match_json_value=lambda o: _has_keys(o, "$escape"),
    match_object=_escape_match_object,
    to_json_value=_escape_to_json,
    from_json_value=_escape_from_json,
)


def _custom_to_json(obj: Any) -> dict[str, Any]:
    return {"$type": obj.type_name(), "$value": obj.to_json_value()}


def _custom_from_json(obj: dict[str, Any]) -> Any:
    type_name = obj["$type"]
    if not has_type(type_name):
        raise ValueError(f"Custom EJSON type {type_name} is not defined")
    return get_factory(type_name)(obj["$value"])


_CUSTOM_CONVERTER = Converter(
    match_json_value=lambda o: _has_keys(o, "$type", "$value"),
    match_object=is_custom_type,
    to_json_value=_custom_to_json,
    from_json_value=_custom_from_json,
)


# Order MUST match the TS ``builtinConverters`` array — Date, RegExp,
# Inf/NaN, Binary, Escape, Custom.
builtin_converters.extend(
    [
        _DATE_CONVERTER,
        _REGEXP_CONVERTER,
        _INFNAN_CONVERTER,
        _BINARY_CONVERTER,
        _ESCAPE_CONVERTER,
        _CUSTOM_CONVERTER,
    ]
)
